tests/test_ledger.py

```diff
@@ -1,445 +1,159 @@
-
-    # Mirror with one single and one collective transaction
-    target = LEDGER_ENTRIES.query("id in [1, 2]")
-    cashctrl.mirror_ledger(target=target, delete=True)
-    expected = cashctrl.standardize_ledger(target)
-    mirrored = cashctrl.ledger()
-    assert txn_to_str(mirrored) == txn_to_str(expected)
-
-    # Mirror with duplicate transactions and delete=False
-    target = pd.concat(
-        [
-            LEDGER_ENTRIES.query("id == 1"),
-            LEDGER_ENTRIES.query("id == 1").assign(id=5),
-            LEDGER_ENTRIES.query("id == 2").assign(id=6),
-            LEDGER_ENTRIES.query("id == 2"),
+    def test_add_ledger_entry(self, ledger, ledger_id):
+        super().test_add_ledger_entry(ledger, ledger_id)
+
+    def test_add_ledger_with_non_existing_vat(self, ledger):
+        # Adding a ledger entry with non existing VAT code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["vat_code"].iat[0] = "Test_Non_Existent_VAT_code"
+        with pytest.raises(ValueError, match="No id found for tax code"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_account(self, ledger):
+        # Adding a ledger entry with non existing account should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["account"] = 33333
+        with pytest.raises(ValueError, match="No id found for account"):
+            ledger.add_ledger_entry(target)
+
+    def test_add_ledger_with_non_existing_currency(self, ledger):
+        # Adding a ledger entry with non existing currency code should raise an error
+        target = self.LEDGER_ENTRIES.query("id == 1").copy()
+        target["currency"] = "Non_Existent_Currency"
+        with pytest.raises(ValueError, match="No id found for currency"):
+            ledger.add_ledger_entry(target)
+
+    @pytest.mark.parametrize("id", [15, 16])
+    def test_adding_transaction_with_two_non_base_currencies_fails(self, ledger, id):
+        # flake8: noqa: E501
+        LEDGER_CSV = """
+            id,     date, account, counter_account, currency,     amount, base_currency_amount, vat_code, text,                             document
+            15, 2024-06-26,     ,            9991,      USD,  100000.00,             90000.00,        ,   Convert 100k USD to EUR @ 0.9375,
+            15, 2024-06-26, 9990,                ,      EUR,   93750.00,             90000.00,        ,   Convert 100k USD to EUR @ 0.9375,
+            16, 2024-06-26,     ,            9991,      USD,  200000.00,            180000.00,        ,   Convert 200k USD to EUR and CHF,
+            16, 2024-06-26, 9990,                ,      EUR,   93750.00,             90000.00,        ,   Convert 200k USD to EUR and CHF,
+            16, 2024-06-26, 9992,                ,      CHF,   90000.00,             90000.00,        ,   Convert 200k USD to EUR and CHF,
+        """
+        # flake8: enable
```

Comment:
Let's use a more compact version to comply with linting rules:

```python
LEDGER_CSV = """
    id,   date, account, counter, currency,    amount, base_amount, text
    0, 2024-06-26,     ,    9991,      USD, 100000.00,    90000.00, Convert USD to EUR
    0, 2024-06-26, 9990,        ,      EUR,  93750.00,    90000.00, Convert USD to EUR
    1, 2024-06-26,     ,    9991,      USD, 200000.00,   180000.00, Convert USD to EUR+CHF
    1, 2024-06-26, 9990,        ,      EUR,  93750.00,    90000.00, Convert USD to EUR+CHF
    1, 2024-06-26, 9992,        ,      CHF,  90000.00,    90000.00, Convert USD to EUR+CHF
"""
```

Instead of:
```python
# flake8: noqa: E501
LEDGER_CSV = """
    id,     date, account, counter_account, currency,     amount, base_currency_amount, vat_code, text,                             document
    15, 2024-06-26,     ,            9991,      USD,  100000.00,             90000.00,        ,   Convert 100k USD to EUR @ 0.9375,
    15, 2024-06-26, 9990,                ,      EUR,   93750.00,             90000.00,        ,   Convert 100k USD to EUR @ 0.9375,
    16, 2024-06-26,     ,            9991,      USD,  200000.00,            180000.00,        ,   Convert 200k USD to EUR and CHF,
    16, 2024-06-26, 9990,                ,      EUR,   93750.00,             90000.00,        ,   Convert 200k USD to EUR and CHF,
    16, 2024-06-26, 9992,                ,      CHF,   90000.00,             90000.00,        ,   Convert 200k USD to EUR and CHF,
"""
# flake8: enable
```